spring-cloud-gcp-autoconfigure/src/test/java/org/springframework/cloud/gcp/autoconfigure/pubsub/it/PubSubChannelAdaptersIT.java

```diff
@@ -0,0 +1,231 @@
+	private PubSubAdmin pubSubAdmin;
+
+	@Autowired
+	private GcpPubSubProperties pubSubProperties;
+
+	@Autowired
+	@Qualifier("inputChannel")
+	private MessageChannel inputChannel;
+
+	@Autowired
+	private PubSubInboundChannelAdapter inboundChannelAdapter;
+
+	@Autowired
+	private PubSubMessageHandler outboundChannelAdapter;
+
+	@BeforeClass
+	public static void checkEmulatorIsRunning() {
+		assumeThat(System.getenv(EMULATOR_HOST_ENVVAR_NAME)).isNotNull();
+	}
+
+	@Before
+	public void setUp() {
+		this.pubSubAdmin.createTopic("desafinado");
+		this.pubSubAdmin.createSubscription("doralice", "desafinado");
+	}
+
+	@After
+	public void tearDown() {
+		this.pubSubAdmin.deleteTopic("desafinado");
+		this.pubSubAdmin.deleteSubscription("doralice");
+	}
+
+	@Test
+	public void sendAndReceiveMessage() {
+		Map<String, Object> headers = new HashMap<>();
+		// Only String values for now..
+		headers.put("storm", "lift your skinny fists");
+		headers.put("static", "lift your skinny fists");
+		headers.put("sleep", "lift your skinny fists");
+
+		this.inputChannel.send(
+				MessageBuilder.createMessage("I am a message.",  new MessageHeaders(headers)));
+
+		Message<?> message = this.channel.receive(5000);
+		assertThat(message).isNotNull();
+		assertThat(message.getPayload()).isInstanceOf(String.class);
+		String payload = (String) message.getPayload();
+		assertThat(payload).isEqualTo("I am a message.");
+
+		assertThat(message.getHeaders().size()).isEqualTo(6);
+		assertThat(message.getHeaders().get("storm")).isEqualTo("lift your skinny fists");
+		assertThat(message.getHeaders().get("static")).isEqualTo("lift your skinny fists");
+		assertThat(message.getHeaders().get("sleep")).isEqualTo("lift your skinny fists");
+	}
+
+	@Test
+	@DirtiesContext
+	public void sendAndReceiveMessageInBytes() {
+		this.inboundChannelAdapter.setMessageConverter(null);
+		this.inputChannel.send(MessageBuilder.withPayload("I am a message.").build());
+
+		Message<?> message = this.channel.receive(5000);
+		assertThat(message).isNotNull();
+		assertThat(message.getPayload()).isInstanceOf(byte[].class);
+		String stringPayload = new String((byte[]) message.getPayload());
+		assertThat(stringPayload).isEqualTo("I am a message.");
+	}
+
+	@Test
+	@DirtiesContext
+	public void sendAndReceiveMessageManualAck() {
+		this.inboundChannelAdapter.setAckMode(AckMode.MANUAL);
+		this.inputChannel.send(MessageBuilder.withPayload("I am a message.").build());
+
+		Message<?> message = this.channel.receive(5000);
+		assertThat(message).isNotNull();
+		AckReplyConsumer acker =
+				(AckReplyConsumer) message.getHeaders().get(GcpHeaders.ACKNOWLEDGEMENT);
+		assertThat(acker).isNotNull();
+		acker.nack();
+		message = this.channel.receive(5000);
+		assertThat(message).isNotNull();
+		acker = (AckReplyConsumer) message.getHeaders().get(GcpHeaders.ACKNOWLEDGEMENT);
+		assertThat(acker).isNotNull();
+		acker.ack();
+		message = this.channel.receive(5000);
+		assertThat(message).isNull();
```

Comment:
Why do we wait here for nothing 5 seconds?

Reply:
Not sure if I understand the question.

In my tests, I had to wait for a little while (e.g., more than one second) for the message to arrive.

In order to make sure that no message arrives, I thought we should wait for as long as the previous calls waited?

Reply:
Well, why should we wait for one more message at all?
Can we just remove these last couple lines at all, otherwise we definitely wait for whole 5 seconds?

As I said: that's very bad to have slow tests and even Stephane has pointed you how to increase performance for Spring Boot tests using slices approach.

Reply:
This method tests that a message which is nacked comes back and when we ack it, it goes away and doesn't come back.
I think it's important to verify that when the message is acked, it doesn't come back.